plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/AbstractBuildBootableJarMojo.java

```diff
@@ -1205,13 +1210,59 @@ private void deploy(List<String> commands) throws MojoExecutionException {
 
     private static void zipServer(Path home, Path contentDir) throws IOException {
         Path target = contentDir.resolve("wildfly.zip");
-        ZipUtils.zip(home, target);
+        zip(home, target);
     }
 
     private void buildJar(Path contentDir, Path jarFile, Artifact artifact) throws MojoExecutionException, IOException {
         Path rtJarFile = resolveArtifact(artifact);
         ZipUtils.unzip(rtJarFile, contentDir);
-        ZipUtils.zip(contentDir, jarFile);
+        zip(contentDir, jarFile);
+    }
+
+    private static void zip(Path contentDir, Path jarFile) throws IOException {
+        try (FileOutputStream fos = new FileOutputStream(jarFile.toFile()); ZipOutputStream zos = new ZipOutputStream(fos)) {
+            Files.walkFileTree(contentDir, EnumSet.of(FileVisitOption.FOLLOW_LINKS), 1,
+                    new SimpleFileVisitor<Path>() {
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                        throws IOException {
+                    if (!contentDir.equals(dir)) {
+                        zip(dir.toFile(), dir.toFile().getName(), zos);
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
+                        throws IOException {
+                    zip(file.toFile(), file.toFile().getName(), zos);
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        }
+    }
+
+    private static void zip(File fileToZip, String fileName, ZipOutputStream zos) throws IOException {
+        if (fileToZip.isDirectory()) {
+            fileName = fileName.endsWith(File.separator) ? fileName.substring(0, fileName.length() - 1) : fileName;
+            fileName = fileName + "/";
```

Comment:
We may want to test this on Windows or do something like:
```
fileName = fileName.replace('\\', '/');
fileName = fileName.endsWith("/") ? fileName : fileName + "/";
```
I don't know for sure, but I wonder if there is a change if someone uses something like `${project.basedir}/some/directory/` on Windows what would happen. I thought it was required for zip entries to use `/` for path separators so I don't recall what happens on Windows if a '\' is used.

Reply:
I guess another option is to use a `Path` instead of a `File` too because you can iterate the names.